### PR TITLE
[Profiling] Query in parallel only if beneficial

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we check index allocation before we do key-value lookups. To reduce latency, key-value lookups are done in parallel for multiple slices of data. However, on nodes with spinning disks, parallel accesses are harmful. Therefore, we check whether any index is allocated either to the warm or cold tier (which are usually on spinning disks) and disable parallel key-value lookups. This has improved latency on the warm tier by about 10% in our experiments.

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
@elasticmachine merge upstream

**Review by rockdaboot — APPROVED** [review]
:+1:

File: x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java [hunk 1]
```diff
@@ -401,9 +401,12 @@ private void retrieveStackTraces(
             return;
         }
         List<String> eventIds = new ArrayList<>(responseBuilder.getStackTraceEvents().keySet());
-        List<List<String>> slicedEventIds = sliced(eventIds, desiredSlices);
         ClusterState clusterState = clusterService.state();
         List<Index> indices = resolver.resolve(clusterState, "profiling-stacktraces", responseBuilder.getStart(), responseBuilder.getEnd());
+        // Avoid parallelism if there is potential we are on spinning disks (frozen tier uses searchable snapshots)
+        int sliceCount = IndexAllocation.isAnyOnWarmOrColdTier(clusterState, indices) ? 1 : desiredSlices;
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java:407]: I just dug through my notes. I believe I did such a test but the superior alternative in all cases was just using a single slice (apparently I did not even bother to write down the results for anything except the default value of `16` and a single slice...).

**Comment by danielmitterdorfer** [comment]
I'm manually backporting this in https://github.com/elastic/elasticsearch/pull/103144.

**Pull request merged**